### PR TITLE
#18133958 user login and authentication

### PR DESCRIPTION
**Pull request by Gabin-ishimwe**
## What does this PR do?
i create a login route for a user and protected some route by adding authentication, and also i made test for this route api
## Description of Task to be completed?
created login route where user will login and give them a token to access other protected pages. this token will be used to authenticate other pages after the user has loged in
## How should this be manually tested?

- clone this repository
- run you server with `yarn run dev` (remember to setup you dev database)
- for testing run your server by `yarn run test` (remember to setup you test database)
- testing these routes in postman: `/api/v1/users/login`
- test this route for authentication: `/api/v1/users/testAuth`
- swagger documentation route: `/api-docs`

## What are the relevant pivotal tracker stories?
[#18133958](https://www.pivotaltracker.com/n/projects/2555280/stories/181339528)
## Screenshots (if appropriate)
image

## Questions:

### CODE REVIEW

**Review by niyonkurueric — APPROVED** [review]

**Review by Code-250 — APPROVED** [review]

File: src/controllers/user.controller.js [hunk 1]
```diff
@@ -26,4 +26,22 @@ export default class UserController {
...
+  async userLogin(req, res, next) {
+    try {
+      const user = await this.userService.userLogin(req.body.email, res);
+      const validation = await comparePassword(req.body.password, user.password)
+      if(validation) {
+        const token = await generateToken({email: user.email}, "1d")
+        res.status(201).header("authenticate", token).json({message: "Logged in successfully", token: token})
+      }
+      else {
+        res.status(401).send("Invalid password")
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: Return json, and change the status code.

File: src/controllers/user.controller.js [hunk 2]
```diff
@@ -26,4 +26,22 @@ export default class UserController {
...
+      }
+    } catch (error) {
+      console.log(error)
+      res.status(404).send({error: error})
```
Comment (ncutixavier) [inline src/controllers/user.controller.js]: - Remove the console and add provide a meaningful message to the user
`res.status(500).send({error: error, message: ""})`
- change status code

File: src/middlewares/passport.middleware.js [hunk 3]
```diff
@@ -1,30 +1,30 @@
...
+// 	new JwtStrategy(
+// 		{
+// 			secretOrKeyProvider: (req, rw, done) => done(null, process.env.SECRETE),
+// 			jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
+// 		},
+// 		(payload, done) => {
+// 			done(null, payload);
+// 		}
+// 	)
+// );
```
Comment (ncutixavier) [inline src/middlewares/passport.middleware.js]: Remove commented code

File: src/middlewares/user.middleware.js [hunk 4]
```diff
@@ -17,3 +19,19 @@ export const checkEmailExist = async (req, res, next) => {
...
+export const authenticateRoute = async (req, res, next) => {
+  try {
+    const token = req.headers.authorization.split(" ")[1];
+    if (token) {
+      const verify = Jwt.verify(token, process.env.SECRETE);
+      req.user = verify
+      next()
+  }
+  } catch (error) {
+    res.status(401).send("Access denied")
```
Comment (ncutixavier) [inline src/middlewares/user.middleware.js]: Return json

File: src/routes/api/user.route.js [hunk 5]
```diff
@@ -16,4 +16,12 @@ routes.post(
...
 );
 
+routes.post('/login', userValidation, async (req, res) => {
+  await new UserController().userLogin(req, res);
+})
+
+routes.get("/testAuth", authenticateRoute, (req, res) => {
+  res.status(200).send("Welcome on the test authentication route")
+})
+
```
Comment (ncutixavier) [inline src/routes/api/user.route.js]: Remove unused code

File: src/services/user.service.js [hunk 6]
```diff
@@ -8,4 +8,18 @@ export default class UserService {
...
+    const userExist = await User.findOne({
+      where: {email: data}
+    })
+
+    if (userExist) {
+      return userExist
+    }
+
+    else {
+      res.status(404).send("User not found")
```
Comment (ncutixavier) [inline src/services/user.service.js]: Return json

**Review by ncutixavier — CHANGES_REQUESTED** [review]

File: src/services/user.service.js (hunk 6 above)
Comment (Gabin-ishimwe) [reply src/services/user.service.js]: done

File: src/routes/api/user.route.js (hunk 5 above)
Comment (Gabin-ishimwe) [reply src/routes/api/user.route.js]: done

File: src/middlewares/user.middleware.js (hunk 4 above)
Comment (Gabin-ishimwe) [reply src/middlewares/user.middleware.js]: done

File: src/middlewares/passport.middleware.js (hunk 3 above)
Comment (Gabin-ishimwe) [reply src/middlewares/passport.middleware.js]: done

File: src/controllers/user.controller.js (hunk 1 above)
Comment (Gabin-ishimwe) [reply src/controllers/user.controller.js]: done

**Comment by ncutixavier** [comment]
Please fulfill your PR Description

**Review by ncutixavier — COMMENTED** [review]
- Add a description of your task in PR Description
- Your deployment is not working

**Review by ncutixavier — APPROVED** [review]

**Pull request merged**